lib/ansible/modules/storage/purestorage/purefa_ds.py

```diff
@@ -0,0 +1,262 @@
+      to right. Each OU should not exceed 64 characters in length.
+  ro_group:
+    description:
+    - Sets the common Name (CN) of the configured directory service group
+      containing users with read-only privileges on the FlashArray. This
+      name should be just the Common Name of the group without the CN=
+      specifier. Common Names should not exceed 64 characters in length.
+  sa_group:
+    description:
+    - Sets the common Name (CN) of the configured directory service group
+      containing administrators with storage-related privileges on the
+      FlashArray. This name should be just the Common Name of the group
+      without the CN= specifier. Common Names should not exceed 64
+      characters in length.
+  aa_group:
+    description:
+    - Sets the common Name (CN) of the directory service group containing
+      administrators with full privileges when managing the FlashArray.
+      The name should be just the Common Name of the group without the
+      CN= specifier. Common Names should not exceed 64 characters in length.
+extends_documentation_fragment:
+- purestorage.fa
+'''
+
+EXAMPLES = r'''
+- name: Delete exisitng directory service
+  purefa_ds:
+    state: absent
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: Create directory service (disabled)
+  purefa_ds:
+    uri: "ldap://lab.purestorage.com"
+    base_dn: "DC=lab,DC=purestorage,DC=com"
+    bind_user: Administrator
+    bind_password: password
+    group_base: "OU=Pure-Admin"
+    ro_group: PureReadOnly
+    sa_group: PureStorage
+    aa_group: PureAdmin
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: Enable exisitng directory service
+  purefa_ds:
+    enable: true
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: Disable exisitng directory service
+  purefa_ds:
+    enable: false
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: Create directory service (enabled)
+  purefa_ds:
+    enable: true
+    uri: "ldap://lab.purestorage.com"
+    base_dn: "DC=lab,DC=purestorage,DC=com"
+    bind_user: Administrator
+    bind_password: password
+    group_base: "OU=Pure-Admin"
+    ro_group: PureReadOnly
+    sa_group: PureStorage
+    aa_group: PureAdmin
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+'''
+
+RETURN = r'''
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pure import get_system, purefa_argument_spec
+
+
+def update_ds(module, array):
+    """Update Directory Service"""
+    changed = False
+    module.exit_json(changed=changed)
+
+
+def enable_ds(module, array):
+    """Enable Directory Service"""
+    try:
+        array.enable_directory_service()
+        changed = True
+    except:
+        changed = False
+    module.fail_json(msg='Enable Directory Service failed: Check configuration')
```

Comment:
I think it should look like this:
``` python
from ansible.module_utils._text import to_native
[...]

def enable_ds(module, array):
    """Enable Directory Service"""
    try:
        array.enable_directory_service()
        changed = True
    except Exception as e:
        module.fail_json(msg='Enable Directory Service failed: Check configuration: %s' % to_native(e))
    module.exit_json(changed=changed)
```

One thing I'm not sure about is how we tell if array.enable_directory_service() made a change or not.  Our users generally define Ansible's idempotence as:
* they describe the state they want in a playbook task
* the module makes the state match what is in the playbook
* the module returns changed=True if something had to be changed to achieve the desired state or changed=False if the state was already as desired.

This code probably does the first two but it's not doing the last one.  I'm not sure how to fix it because I don't know what array.enable_directory_service() does.  If it returns a specifc exception if the directory_service was already enabled, then we catch that specific exception and set changed=True for that.  If it returns a status code telling whether it had to make a change or not, then we can use that to set changed.  If neither of those, we would have to query the current state of the toggle first.  If the toggle is already enabled, then set changed=False and call module.exit_json().  If the toggle is not yet enabled, call array.enable_directory_service(), set changed=True, and then call module.exit_json().

(Also note, these notes on changes also apply to the other verbs you have in this module: disable_ds, delete_ds, and create_ds)

Reply:
This is all checked for in the main() function. It can't get into these functions if the DS isn't already in the correct state for these functions to operate on.